### DISCUSSION TITLE
Calculate the jacobian of the end effector

### DISCUSSION THREAD

**Discussion by blooop** [NVlabs/curobo#150] [Software Q&A] [answered]
Hi,

I am interested in using curobo the calculate the jacobian of the end effector so I can work out the manipulability ellipsoid etc. 

I added these lines to the bottom of the ik example:
https://curobo.org/source/getting_started/2a_python_examples.html#inverse-kinematics

```
from curobo.cuda_robot_model.cuda_robot_model import CudaRobotModel
kin_model = CudaRobotModel(robot_cfg.kinematics)
print(kin_model.get_state(q_solution,calculate_jacobian=True))
```

In the code and docs there are several references to jacobians not being supported yet, and one of those errors happens if I try to set calculate_jacobian=True.

```
  File "/usr/local/lib/python3.10/dist-packages/curobo/cuda_robot_model/cuda_robot_model.py", line 243, in get_state
    out = self.forward(q, link_name, calculate_jacobian)
  File "/usr/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/usr/local/lib/python3.10/dist-packages/curobo/cuda_robot_model/cuda_robot_model.py", line 231, in forward
    raise NotImplementedError
NotImplementedError
```

I was wondering if there was another way to get the jacobian, as it looks like jacobians are used elsewhere, e.g.
https://github.com/NVlabs/curobo/blob/88eac64edca2d2b641c1622e63f3e750157596c4/src/curobo/rollout/dynamics_model/kinematic_model.py#L475

Thanks

**Comment by balakumar-s** [comment, marked as answer]
We never got around to implementing Jacobian in our kinematics kernel. We wanted to support it at some point in the future, hence some code snippets exist in the current code. We can try to implement a non-optimized version of Jacobian first and push it to the code base. 

Would you need to back-prop through the Jacobian? back-prop is where it might be tricky to implement.

**Reply by blooop** [reply]
Thanks.  I don't need backprop and have a workaround using pinocchio to calculate it for the moment.  It's not critical I was just wondering in case there was a solution that was already mostly there.

**Reply by balakumar-s** [reply]
It took us a while to efficiently implement the Jacobian. It's now available. Check: https://nvlabs.github.io/curobo/latest/getting-started/forward_kinematics.html
